### PR TITLE
[dialogs] Busy Dialog Improvements

### LINKED ISSUE #14838: Leia addon crash on xbmc.executebuiltin(PlayMedia...
### Describe the bug

Windows Runtime Error after an addon executes 
`xbmc.executebuiltin('PlayMedia("plugin://plugin.video.crackle/?id=2537&mode=103&type=movies")')`

Also, ActivateWindow does nothing but should launch into a video window.

## Expected Behavior

A movie is played from the Crackle plugin.video.crackle addon.

## Actual Behavior

Windows crashes with a Runtime Error.

## Possible Fix

Kodi Core revert to Krypton behaviour.

### To Reproduce

Install plugin.program.simple.favourites from this zip.
https://github.com/primaeval/repository.primaeval/raw/master/jarvis/plugin.program.simple.favourites/plugin.program.simple.favourites-0.1.13.zip
Install Crackle from Kodi repo.
Kodi\Addons\Programs\Simple Favourites\Add\Video Addons\Crackle\Movies
"Add" a movie from Context Menu.
Back out of Addon.
Go back in to Simple Favourites.
Click on the movie.
Kodi crashes with a Runtime Error.

The xbmc.executebuiltin(PlayMedia command is launched from here:
https://github.com/primaeval/plugin.program.simple.favourites/blob/master/main.py#L95

### Debuglog

The debuglog can be found here:
https://paste.kodi.tv/bihoqubali.kodi

### Screenshots 
Here are some links or screenshots to help explain the problem:

error

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

Screenshot after clicking on movie from top of Simple Favourites addon.

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [x] Linux
 - [ ] OSX
 - [ ] Raspberri-Pi
 - [x] Windows
 - [ ] Windows UWP

 - Operating system version/name:
 - Kodi version:
Kodi (18.0-RC1 Git:20181109-a948f998d9). Platform: Windows NT x86 64-bit
and
Kodi (18.0-BETA3 Git:18.0b3-Leia). Platform: Linux x86 64-bit
Running on LibreELEC (official): 8.90.006 9.0, kernel: Linux x86 64-bit version 4.18.11

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### LINKED ISSUE #16191: Kodi crash if addon doesnt provide stream
### Describe the bug
Here is a clear and concise description of what the problem is:

If switching in "Radio.de" Addon to a channel which doesnt provide a stream Kodi crashes.

(Instead of stopping the busy spinner.)

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

No crash if a Radio stream isnt available

## Actual Behavior

it crashs

### To Reproduce

1.  Install the "Radio.de" Addon from the official Repository.
2.  Open the Radio Addon.
3.  (browse for a Stream which isnt available, below as example: )
4.  within the Addon go to "Nach Musikrichtungen" (browser by genre).
5. Choose "Jazz"
6. Try to play one of the "1Jazz" Channels (by clickin on one) -> "1Jazz- Bebop" (23th listitem - sortby: default)
7. Crash

### Debuglog

The debuglog can be found here: https://paste.ubuntu.com/p/XX53FxyHKq/

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

Just would be nice if i could click an a Radio Addon Channel without have to fear that Kodi crashes, while browsing thrue the cool world of onlineradio.

I knew, the Addon wasnt updated since a while, but still would be nice if Kodi wouldnt crash. would be enought if it just stops to try to play such channels.

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [x] Windows
 - [ ] Windows UWP

 - Operating system version/name: Win 10
 - Kodi version: 18.2

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### LINKED ISSUE #16756: Kodi crashes when 2 concurrent busy dialogs occur
### Describe the bug

Kodi crashes while browsing in the official YouTube add-on if 2 "busy/wait" actions occurs at once.

## Expected Behavior

Kodi does not crash.

## Actual Behavior

Kodi crashes.

### To Reproduce

1. Add 2 videos from YouTube to the playing queue (use the "Queue video" action).
2. Go back to the GUI and leave the first video playing in the background.
3. Execute just about any other "busy" action in the add-on (e.g. a search for another video) just before the first video ends and the second one tries to start in the background (so that both the second video loading and the search occurs at once).
4. Kodi crashes.

### Debuglog

The debuglog can be found here:
http://ix.io/1YCl
The crashlog is from Amlogic MX2 box running Kodi 18.2 because that is the slowest device laying around right now but the same happens on at least 2 different devices as well.

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)

The timing can be a little tricky but it is more easy to reproduce on a slow device (e.g. RPi) where the loading times are bigger.

It is probably not YouTube specific. At least I've tested that the second action can be executed from another add-on while the YouTube video is loading in the background and the Kodi still crashes.

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [X] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: LibreELEC 9.1.501 (tested on Rockchip and Amlogic)
 - Kodi version: 18.4/Leia

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### LINKED ISSUE #20297: Busy dialog + kodi crashes
### Describe the bug

This particular issue can be reproduced if you are using the plugin.program.autocompletion v2.0.2 from my repo (or the official repo) and appears to be a timing issue with the busy dialog and a jsonprc call for "Input.SendText"
If you dont close the busy dialog before you try and send the text to the input box kodi will completely crash, with no errors to the log, it just stops:

On android it requires a 250ms pause between closing the busy dialog and then the Input.SendText event, without this pause kodi will consistently crash every time you select an autocomplete option.

## Expected Behavior
The busy dialog should close on its own when an event occurs which would otherwise be blocked by the busy dialog.

Kodi should not completely crash without any particular messages to the log to describe what the actuall error which occurred was.

## Actual Behavior
Selecting the autocomplete option causes the busy dialog to load, if you try and send the inputsendtext event while the busy dialog is on screen kodi crashes.
If you close the busy dialog first, but dont have a sufficiently long pause before the input send text event, kodi crashes.

The busy dialog also casues issues for me with youtube, videos load the busy dialog right before they start playing. It will close when the video starts up, but just having the busy dialog open prior to "creating demuxer" seems to cause the video load in a small section of the screen, it will then go full screen but the video will then be very jerky and buffer contstantly.
Tailing the log and forcing the busy dialog to close and fullscreenvideo to activate immediately as soon as "VideoPlayer::OpenFile:" appears in the logs seems to stop that particular issue, but again if the busy dialog is a blocker to fullscreenvideo opening, it should close automatically first and then the video should start playing.

## Possible Fix
Make the busy dialog event less blocking?
Have kodi more reliably close the busy dialog and verify its fully closed before it starts the next action?

Maybe have the busy dialog close and reopen when other events happen in kodi?
But the busy dialog can and will cause kodi to crash in strange ways so perhaps it needs completely rebuilt in a new way so that it closes more reliably.

### To Reproduce
Using plugin.program.autocompletion on android v2.0.1/2.0.2 from the kodi repo, start typing, select an autocomplete option, kodi crashes.
If there are xbmc.sleep pauses in "~/.kodi/addons/plugin.program.autocompletion/plugin.py" after the "xbmc.executebuiltin('Dialog.Close(busydialognocancel)')", remove the pauses and it will consistently crash every time.

Seems to be similar to this problem:
https://github.com/xbmc/xbmc/issues/19304
(Note the youtube mention in that issue)

### Debuglog
(logs dont help because it crashes before anything is logged).
https://github.com/henryjfry/repository.thenewdiamond/files/7319133/kodi.log

### Your Environment

 - [x ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] OSX
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name:
 - Kodi version:
Kodi 19.2

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### LINKED ISSUE #20468: Kodi crashes if a youtube video is shared through Kore while the previous video is still in the throbber stage
### Describe the bug
Here is a clear and concise description of what the problem is:

When you share a youtube video to kodi to be played by Plugin for Youtube, sometimes the throbber loading screen takes a while longer. If you share another video to be played or queued while this is happening, whole kodi will crash

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

No crash should happen

## Actual Behavior

Kodi should abandon the previous video and play the new one or put the new video at the back of the queue respectively

### To Reproduce

1. Find a youtube video that takes a longer time to buffer on Kodi
2. Select to share a video to Kodi through Kore
3. Take any other video and share it to Kodi

### Debuglog

The debuglog can be found here:
[kodi_crashlog-20211007_141534.log](https://github.com/xbmc/xbmc/files/7482113/kodi_crashlog-20211007_141534.log)

### Screenshots 
Here are some links or screenshots to help explain the problem:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] OSX
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Arch Linux
 - Kodi version: 19.3.0

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by ksooo**
A couple of busy dialog related cleanups and improvements:

* Remove dead progress control code from `CGUIDialogBusy` and `CGUIDialogBusyNoCancel`. The progress was never set, `CGUIDialogNoCancel` did not even have a method to set it! . The code that set progress was removed years ago and it seems it was simply forgotten to cleanup the dialog code.
* Cleanup `CUPnPPlayer` code. remove unneeded calls to close the dialog as this gets done automatically by `CGUIDialogBusy::WaitOnEvent`. Could be fallout from refactroings done years ago.
* Simplify `CPowerMananager` code to use the simple `CGUIDialogNoCancel` as power management actions cannot be cancelled.
* **Most important commit in this PR**: Make `CGUIDialogBusy::WaitOnEvent` reentrant instead of throwing an exception in case this method gets called when another busy dialog is already running.
* Prevent blocking of GUI when opening internet streams. Put long lasting operation to create and load the playlist into a thread and display a busy dialog until create/load has finished. Note: what actually takes long here is mime type detection via HAED (?) HTTP request.
*  Prevent blocking of GUI when opening internet streams. Put mime type detection into VideoPlayer's worker thread instead of doing this in `CVideoPlayer::OpenFile`which gets called from the GUI thread.

Changes were carefully runtime-tested on macOS and Android, latest master.

No idea who could review the changes as many and partly orphaned components are affected.

### CODE REVIEW

File: xbmc/dialogs/GUIDialogBusy.h
Comment (enen92) [inline xbmc/dialogs/GUIDialogBusy.h:51]: shouldn't this be an atomic integer instead?

File: xbmc/dialogs/GUIDialogBusy.cpp
Comment (enen92) [inline xbmc/dialogs/GUIDialogBusy.cpp:79]: Since the windowmanager creates and adds this window on startup and only a single instance exists throughout the whole app lifecycle, wouldn't it make more sense for this window to be a singleton? It just looks odd that this class defines the windowmanager as a friend class and then calls the window manager to get itself

File: xbmc/dialogs/GUIDialogBusy.h
Comment (ksooo) [reply xbmc/dialogs/GUIDialogBusy.h:51]: I don't think so. `CGUIDialogBusy::WaitOnEvent` must be called exclusively from the rendering (the main) thread.

**Review by enen92 — COMMENTED** [review]
Thanks a lot for this.
I've runtime tested the PR and didn't find any issues. I think this will solve a lot of crashes we currently have with addons.
I also think this is a better approach than the "user-initiated/system-initiated" and disabling of interfaces approach - making it re-entrant and working just like a queue of tasks goes in line with what I initially though for this. Doesn't solve one of the main design issues (the spreading of gui stuff everywhere) but certainly makes it a step closer. +1 from me, left a few comments

File: xbmc/Application.cpp
Comment (ksooo) [reply xbmc/Application.cpp:2130]: Why should it be moved? This class is an implementation detail of `CApplication`, not used elsewhere. If we detect scenarios where this class could be reused, then it's time to move it out of this cpp file, yes. But not now.

**Comment by Ch1llb0** [comment]
Would this also warrant a backport to v19 as it's a bug fix?

**Comment by ksooo** [comment]
> Would this also warrant a backport to v19 as it's a bug fix?

No, regression risk is imo too high.

File: xbmc/dialogs/GUIDialogBusy.h
Comment (enen92) [reply xbmc/dialogs/GUIDialogBusy.h:51]: Yeah you're right. But this is not that obvious any way it can be made safer? Maybe throwing a logic error if `waitOnEvent` is called out of the rendering thread (ref https://github.com/xbmc/xbmc/pull/21441)?

File: xbmc/dialogs/GUIDialogBusy.cpp
Comment (enen92) [reply xbmc/dialogs/GUIDialogBusy.cpp:79]: > I thought it is more clever/safe to make the ctor/dtor private and add window manager as friend. If this is not acceptable, I could make the ctor/dtor public again.

No, maybe i did not express myself correctly. This is fine.
I know the call to `WindowManager` to get the busydialog was already there and it exists because the method is static. 
I was trying to say that the simple fact this call exists is a sign the design is a bit fragile. I mean, it doesn't make much sense to be the dialog asking the windowmanager (responsible for its creation) to get a pointer to itself so that some logic take place. Maybe this is something that can be improved later but I also don't know how to improve this at this point.

File: xbmc/Application.cpp
Comment (enen92) [reply xbmc/Application.cpp:2130]: No technical reason, just because `3511` lines of code in CApp is probably enough already :)

File: xbmc/dialogs/GUIDialogBusy.h
Comment (ksooo) [reply xbmc/dialogs/GUIDialogBusy.h:51]: I actually used exactly this safety net (exception) while doing runtime tests for this change, but removed it because I needed to include Application.h for this to check the current thread. If this new dependency would be okay I could re-introduce the check/exception.

File: xbmc/dialogs/GUIDialogBusy.cpp
Comment (ksooo) [reply xbmc/dialogs/GUIDialogBusy.cpp:79]: Then let’s keep it as it is for now. Code quality did not decrease, so it should be okay.

File: xbmc/Application.cpp
Comment (ksooo) [reply xbmc/Application.cpp:2130]: I think let’s keep it there for now. A todo comment is imo not needed.

**Review by enen92 — APPROVED** [review]
Adding a green from my side.
List of issues this will fix: https://github.com/xbmc/xbmc/issues?q=is%3Aissue+is%3Aopen+busydialog

**Comment by Ch1llb0** [comment]
> No, regression risk is imo too high.

The crashes themselves are such a big issue themselves, wouldn't it be worth testing at least?

**Comment by ksooo** [comment]
> The crashes are such a big issue themselves, wouldn't it be worth trying and testing at least?

We live with those crashes since 2018. If it is such a pain for you, you could try a recent Nexus nightly. Pretty stable, if you ask me.

**Comment by ksooo** [comment]
Heads up: I will merge this  tomorrow if nobody objects.

File: xbmc/dialogs/GUIDialogBusy.h
Comment (FernetMenta) [reply xbmc/dialogs/GUIDialogBusy.h:51]: > CGUIDialogBusy::WaitOnEvent must be called exclusively from the rendering
Where does this info come from?

https://github.com/xbmc/xbmc/blob/9f699af44bde1b8b3448c3d4162fbb089974d38e/xbmc/pvr/windows/GUIWindowPVRGuide.cpp#L887

File: xbmc/dialogs/GUIDialogBusy.cpp
Comment (FernetMenta) [inline xbmc/dialogs/GUIDialogBusy.cpp:98]: Consider a boarder case where CGUIDialogBusy::WaitOnEvent is called a thousand times from components running on the main thread. A thousand waiters will call ProcessRenderLoop and and each one will render the GUI and drive the process loop. This will obviously not work. (it won't even work correctly if there would be only 2 waiters)

Comment (ksooo) [reply xbmc/dialogs/GUIDialogBusy.cpp:98]: > A thousand waiters will call ProcessRenderLoop and and each one will render the GUI and drive the process loop. 

I do not see the problem here. There is only one active "ProcessRenderLoop loop" at a time, the one from the top-most waiter on the stack, all others are just waiting for their call to ProcessRenderLoop to return.

Comment (ksooo) [reply xbmc/dialogs/GUIDialogBusy.cpp:98]: CBusyDialog::WaitOnEvent -> dialog->ProcessRenderLoop(false); (waits for ProcessRenderLoop to return) 
-> CBusyDialog::WaitOnEvent -> dialog->ProcessRenderLoop(false); (and wait for ProcessRenderLoop to return)-> l
-> CBusyDialog::WaitOnEvent ->dialog->ProcessRenderLoop(false); (rendering in a loop until cancelled or event was set)

Comment (FernetMenta) [reply xbmc/dialogs/GUIDialogBusy.cpp:98]: > all others are just waiting for their call to ProcessRenderLoop to return.

where do you think they are waiting?

```
bool CGUIWindowManager::ProcessRenderLoop(bool renderOnly)
{
  bool renderGui = true;

  if (CServiceBroker::GetAppMessenger()->IsProcessThread() && m_pCallback)
  {
    renderGui = m_pCallback->GetRenderGUI();
    m_iNested++;
    if (!renderOnly)
      m_pCallback->Process();
    m_pCallback->FrameMove(!renderOnly);
    m_pCallback->Render();
    m_iNested--;
  }
  if (g_application.m_bStop || !renderGui)
    return false;
  else
    return true;
}
```

The top most waiter calls m_pCallback->Process(), which may call a function that calls ProcessRenderLoop again. This will end up having multiple render loops on the stack. This was the reason for making the 2nd call bail out.

Comment (FernetMenta) [reply xbmc/dialogs/GUIDialogBusy.cpp:98]: @ksooo some years ago I spent quite some time on investigating hard to debug issues. I found that having multiple render loops on the stack lead to unpredictable behaviour even cycles after the stack got cleared. Now as you set it back to the state it was before, the issues will come back as well.

Comment (ksooo) [reply xbmc/dialogs/GUIDialogBusy.cpp:98]: But what exactly was that unpredictable behavior? Back then when you introduced the exception I was asking for an explanation: https://github.com/xbmc/xbmc/pull/13833#issuecomment-543170237

I think I understand what you ment and think my PR fixes the „not cancallable“ problem you mentioned. 

On first Open, the window gets registered in the map of active windows at window manager
On second Open, nothing will be inserted in that map, because the key is the window id and from the first call there is already an entry.
On Close from the logical second dialog, the entry will be removed from the window mgr map
And that is the actual problem, because from now on no actions will be dispatched to the logical first dialog (as this is done by the window mgr via the map of active windows
This is what fernet means with “cannot be cancelled” - abort action comes not through to the first busy dialog
My solution is simply not to call open/close if there is already a busy dialog active, but rather to jump straight into the renderlopp-loop, logically exactly what you did to solve the problem in cmediawindow.

**Comment by FernetMenta** [comment]
For the record: 
The busy dialog signals to the user that their action is still running. Since there can only be a single user initiated action at a given time, calling CGUIDialogBusy::WaitOnEvent while the busy dialog is running is a logical error. Long running functions on the main thread abused the busy dialog to keep the main loop going. Means CGUIDialogBusy::WaitOnEvent was called by functions on lower layers.
What this PR calls an improvement is mainly a revert of some code that should have helped fixing the root cause.

File: xbmc/dialogs/GUIDialogBusy.cpp
Comment (FernetMenta) [reply xbmc/dialogs/GUIDialogBusy.cpp:98]: The problem is not the dialog itself. The problem is the render loop. If you have more than a single instance of CGUIWindowManager::ProcessRenderLoop on the stack, the strict ordering of FrameMove, Render gets broken because one of those methods may call the BusyDialog again.

**Comment by ksooo** [comment]
Seems the idea that the exception helps to find and fix the  logical errors did not work. Since you introduced the exception I have not seen a single PR with a fix, but a lot of crash reports caused by that exception. Since nobody seems to be out there currently with the skills and mood to fix the actual problems, I consider your idea turned out to make things worse from users perspective and this since 2019 where you have introduced the exception. 

Henning that said, under the given circumstances, it seems right to merge this PR.

**Comment by enen92** [comment]
It's quite hard to draw a line on the concept of user initiated actions, the busydialog just gets triggered from everywhere to signal Kodi is busy by now... 
For instance if you play music via the GUI (play from here) and initiate a playlist, the busydialog will get activated whenever Kodi queues a new song. Is it a user initiated action?
User triggers a slow python script (thus user initiated) that script for some reason triggers another one, is the second a user initiated action?
Not saying this won't cause further problems but maybe the real root case of this case is actually the way the busydialog is perceived, at a time it has maybe diverged too much from the original intention.
I'd rather see it (or better some busy handler) as a queue of tasks. You should be able to queue as much tasks as you want, as long as you can still cancel/or notify cancellation to them all. The busy animation should not be linked to a single event/task but rather on the fact the queue is empty.

**Comment by ksooo** [comment]
Okay, no compromises then. I only hope somebody steps up and implements the right solution for Nexus or even something we can backport.

**Comment by FernetMenta** [comment]
> For instance if you play music via the GUI (play from here) and initiate a playlist, the busydialog will get activated whenever Kodi queues a new song. Is it a user initiated action?

@enen92 it doesn't matter because this is not the root cause. Kodi has implemented an ancient design pattern called main loop (this is what you got 30 years ago when you created a new project in Visual Studio) 
https://github.com/xbmc/xbmc/blob/master/xbmc/Application.cpp#L1842
The main loop handles user io as well as rendering. If you block this main loop, the app is frozen. The BusyDialog is a hack that keeps the main loop going by driving it from another location: https://github.com/xbmc/xbmc/blob/e22e7394f3f648323ca7543939dfc40fa7015aa8/xbmc/guilib/GUIWindowManager.cpp#L1386

btw: this main loop anti-pattern is the source of many other issues

**Comment by enen92** [comment]
Yeap, unfortunately I got to the same conclusion yesterday. 
This will be pretty much impossible to solve without either splitting the threads or avoiding cases with multiple busydialogs being triggered at the same time. The last, due to the way Kodi works - especially with builtins where you have the same point of entrance for pretty much all code paths - skins, scripts and core - is nearly impossible to make it fail proof.
Unfortunately I think we'll have to live with the crashes for yet a few more years to come

**Comment by KOPRajs** [comment]
It is 2024, Omega is out, yet the crashes are still ruining my Kodi experience.

Is there still no hope to fix this? How about at least canceling the second busy action (let it fail) instead of letting Kodi to crash? Or even blocking the main thread until the second busy action is finished seems better than a crash. More ideas?

**Comment by garbear** [comment]
Original fix split out to https://github.com/xbmc/xbmc/pull/27038 and updated to fix stacking of render loops.

**Pull request closed without merging**
